### DISCUSSION TITLE
Handle empty string as undefined

### DISCUSSION THREAD

**Discussion by Jancera** [react-hook-form/react-hook-form#12411] [Q&A]
# Problem

In my project, I need to remove all empty strings from form results. For example, if a user types something and then deletes it, that empty string should be converted to `undefined`.

To do this, I decided to update my input component, which uses the Controller strategy, to check if the value is an empty string and, if so, set it to `undefined`.

```typescript
export default function FormInput({ name, label }: InputProps) {
  const { control } = useFormContext();

  const handleValue = (value: string) => {
    if (value === "" || value === null) {
      return undefined;
    }

    return value;
  };

  return (
    <Controller
      name={name}
      control={control}
      render={({ field, fieldState: { error } }) => {
        return (
          <div>
            {label ? <label>{label}</label> : null} <br />
            <input
              id={name}
              name={field.name}
              value={field.value}
              onChange={(e) => {
                const handledValue = handleValue(e.target.value);
                field.onChange(handledValue);
              }}
              onBlur={() => {
                field.onBlur();
              }}
            />
            {error ? <p>{error.message}</p> : null}
          </div>
        );
      }}
    />
  );
}
``` 

That was working well, until an specific implementation of the forms into the project. We have some forms divided into steps, so those forms are using the `FormProvider` to read/store values. I have a conditional rendering logic to decide what step are active and the things starts to get weird when I change to the second step. If I delete a input value to be empty string, my component will change it to `undefined` and set this value for that field on the context, but visually that input will have its value equal to the initial value that was set to that field when the component was rendered.

I am not using the `defaultValues` option of the `useForm` hook, when my component renders I use `setValue` to set each field value to the initial value that I want, this way I could avoid the behavior of setting undefined for the field value and the `defaultValue` be applied. 

```typescript
function App() {
  const [step, setStep] = useState(0);

  const form = useForm({
    mode: "onChange",
  });

  useEffect(() => {
    form.setValue("name", "nameDefault");
    form.setValue("age", "10");
    form.setValue("city", "cityDefault");
  }, [form]);

  const handleNext = () => {
    setStep(step + 1);
  };

  const handleBack = () => {
    setStep(step - 1);
  };

  return (
    <div>
      <form onSubmit={(d) => console.log(d)}>
        <FormProvider {...form}>
          {step === 0 ? <StepOne /> : <StepTwo />}
        </FormProvider>

        <div>
          <button onClick={handleBack} disabled={step === 0}>
            Back
          </button>
          <button onClick={handleNext} disabled={step === 1}>
            Next
          </button>
          <input type="submit" value="Submit" />
        </div>
      </form>

      <DevTool control={form.control} />
    </div>
  );
}
``` 

I wonder why the form context value are correct and on the screen the value switches from empty string to the input initial value and also why this only happens after changing the first step.

# Repository

[I have a GitHub repo with a sample project so you guys can see what is happening](https://github.com/Jancera/rhf-bug-sample)

# Running the project

- Clone the repository
- Run `npm install`
- Run `npm run dev`

# Replicate the problem

- Click the `Next` button
- Delete the value inside `City` input
- When you delete the last letter, the value will be reseted to "cityDefault", but the submit button prints the city value as `undefined`

**Comment by omergungor11** [comment]
Keep `''` in the controlled input, and convert it to `undefined` at the submission boundary. In the linked sample, `handleValue` sends `undefined` to `field.onChange`, which is outside the supported `Controller` contract ([docs](https://react-hook-form.com/docs/usecontroller/controller)). Passing an undefined `value` also breaks the input's consistently controlled state ([React docs](https://react.dev/reference/react-dom/components/input#im-getting-an-error-a-component-is-changing-an-uncontrolled-input-to-be-controlled)).

For your `FormInput`, remove `handleValue` and pass the field handlers through:

```tsx
<Controller
  name={name}
  control={control}
  render={({ field, fieldState: { error } }) => (
    <div>
      <label htmlFor={name}>{label}</label>
      <input id={name} {...field} value={field.value ?? ''} />
      {error && <p>{error.message}</p>}
    </div>
  )}
/>
```

Initialize the sample's values in `useForm`, removing the initialization effect:

```tsx
const form = useForm({
  mode: 'onChange',
  defaultValues: {
    name: 'nameDefault',
    age: '10',
    city: 'cityDefault',
  },
  shouldUnregister: false, // retain fields while their step is unmounted
});

const onSubmit = form.handleSubmit(values => {
  const payload = Object.fromEntries(
    Object.entries(values).map(([key, value]) => [
      key,
      value === '' ? undefined : value,
    ]),
  );
  console.log(payload);
});

// <form onSubmit={onSubmit}> ... </form>
```

That conversion covers the flat object of string fields in your reproduction. Nested objects/arrays would need a schema-aware conversion. Also, `undefined` isn't a JSON value: `JSON.stringify` omits object properties with that value. If the API needs an explicit instruction to clear an existing value, use whatever representation it defines (often `null`).

One separate fix in the linked `App.tsx`: give both Back and Next `type="button"`, otherwise they also submit the form.

I verified the corrected pattern with RHF 7.88.0 / React 19.2.4 in a React Testing Library + jsdom test under StrictMode: clear City, go Back, go Next, and it stays empty; submission produces `city: undefined`, nonempty values remain intact, and navigation doesn't submit. This addresses the unsupported empty-value path without relying on the particular fallback behavior of the older version in the sample.

**Comment by burakcankrpnr** [comment]
I think the key issue here is that there are two different concerns being mixed together: the value required by the controlled input and the value required by the API/submission payload.

For a controlled input, an empty field should remain `""`. Converting it to `undefined` inside `field.onChange()` makes the input transition away from a normal controlled string value, which is why the UI can fall back to the previously registered/default value when the step is mounted again.

I would keep the form state UI-friendly:

```tsx
onChange={(e) => field.onChange(e.target.value)}

const normalizedValues = Object.fromEntries(
  Object.entries(values).map(([key, value]) => [
    key,
    value === "" ? undefined : value,
  ])
);
